### PR TITLE
add reduce arg to PoissonNLLLoss

### PR DESCRIPTION
**Pull request by kevinzakka**
As per #264. When reduce is False, PoissonNLLLoss outputs a loss per element of the input tensor. When reduce is True (default), the current behavior is kept.

This did not require changing any C or CUDA files as PoissonNLLLoss is implemented purely in python.

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1130,10 +1134,13 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Tru
     if full:
         mask = target > 1
         loss[mask] += (target * torch.log(target) - target + 0.5 * torch.log(2 * math.pi * target))[mask]
-    if size_average:
-        return torch.mean(loss)
+    if not reduce:
+      return loss
```
Comment (zou3519) [inline torch/nn/functional.py]: I think the indentation is a little off here -- it should be 4 spaces

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1130,10 +1134,13 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Tru
...
     else:
-        return torch.sum(loss)
+      if size_average:
+          return torch.mean(loss)
+      else:
+          return torch.sum(loss)
```
Comment (zou3519) [inline torch/nn/functional.py]: nit: It's a little cleaner if you get rid of the elses, ie:
```
if not reduce:
    return loss
if size_average:
    ...
```
but this might just be my personal preference.

File: test/test_nn.py [hunk 3]
```diff
@@ -3728,6 +3728,13 @@ def forward(self, *args):
         module_name='PoissonNLLLoss',
         input_size=(2, 3, 4, 5),
         target_fn=lambda: torch.randn(2, 3, 4, 5).floor_().abs_(),
+        desc='non_full_loss',
```
Comment (zou3519) [inline test/test_nn.py]: To be consistent with the other loss function tests, the desc should be 'no_reduce', so the full test name is `test_PoissonNLLLoss_no_reduce`

File: test/test_nn.py [hunk 4]
```diff
@@ -3728,6 +3728,13 @@ def forward(self, *args):
...
+    ),
+    dict(
+        module_name='PoissonNLLLoss',
+        constructor_args=(True, False, True, 1e-8, False),
+        input_size=(2, 3, 4, 5),
+        target_fn=lambda: torch.randn(2, 3, 4, 5).floor_().abs_(),
```
Comment (zou3519) [inline test/test_nn.py]: Could you add a `reference_fn` for this and the `no_reduce` (`non_full_loss`) case? See `kldivloss_no_reduce_test` [here](https://github.com/pytorch/pytorch/pull/3330/files#diff-d89baec73022f5f511c5beb5ce6498dfR3677) for an example

**Review by zou3519 — COMMENTED** [review]
Thanks for tackling this, @kevinzakka. This looks great! I had a few minor comments

File: test/test_nn.py [hunk 5]
```diff
@@ -3728,18 +3728,30 @@ def forward(self, *args):
...
 
 
+def poissonnllloss_no_reduce_test():
+    t = Variable(torch.randn(10, 10))
+    return dict(
+        fullname='PoissonNLLLLoss_no_reduce',
+        constructor=wrap_functional(
+            lambda i: F.poisson_nll_loss(i, t.type_as(i), reduce=False)),
+        input_fn=lambda: torch.rand(10, 10),
+        reference_fn=lambda i, _:
```
Comment (zou3519) [inline test/test_nn.py]: No reference function needed anymore :)

File: torch/nn/functional.py [hunk 6]
```diff
@@ -1130,10 +1134,11 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Tru
     if full:
         mask = target > 1
         loss[mask] += (target * torch.log(target) - target + 0.5 * torch.log(2 * math.pi * target))[mask]
+    if not reduce:
+      return torch.mean(loss, 1)
```
Comment (zou3519) [inline torch/nn/functional.py]: This should just be `return loss`.

I know the documentation for reduce says "loss per batch element" but the behavior of reduce is really putting out a loss per element of the input

File: test/common_nn.py [hunk 7]
```diff
@@ -245,6 +245,10 @@ def get_weight(m):
     ),
 ]
 
+def poissonnllloss_reference(input, target, log_input=True, full=False,
```
Comment (zou3519) [inline test/common_nn.py]: Because PoissonNLLLoss is implemented in python, it doesn't make sense to have a reference function for it -- good catch and my apologies! Please remove this

**Review by zou3519 — COMMENTED** [review]
Good catch - the reference function isn't necessary because PoissonNLLLoss is written in Python. One more minor comment below.

The test seems to be failing on the CI. You can run it directly with
`python test/test_nn.py TestNN.test_poissonnllloss_no_reduce` and see what's up.

File: test/test_nn.py [hunk 8]
```diff
@@ -3728,18 +3728,28 @@ def forward(self, *args):
         module_name='PoissonNLLLoss',
         input_size=(2, 3, 4, 5),
         target_fn=lambda: torch.randn(2, 3, 4, 5).floor_().abs_(),
-        desc='reduced_loss',
+        desc='no_full_loss', # without sterling approx
     ),
     dict(
         module_name='PoissonNLLLoss',
-        constructor_args=(False, True, True),
+        constructor_args=(False, True, True, 1e-8, True),
```
Comment (zou3519) [inline test/test_nn.py]: Nit: This didn't need to be changed (since both 1e-8 and reduce=True are default arguments)

File: torch/nn/functional.py [hunk 9]
```diff
@@ -1122,6 +1122,10 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Tru
             the losses are instead summed for each minibatch. Default: ``True``
         eps (float, optional): Small value to avoid evaluation of log(0) when
             log_input=False. Default: 1e-8
+        reduce (bool, optional): By default, the losses are averaged
+            over observations for each minibatch, or summed, depending on
+            size_average. When reduce is False, returns a loss per batch
```
Comment (zou3519) [inline torch/nn/functional.py]: Nit: False should be surrounded with `` like True

**Review by zou3519 — APPROVED** [review]
Looks good to me! Thanks @kevinzakka 

Two very minor comments and it'll be good to go :)

**Comment by kevinzakka** [comment]
Done, thanks tons @zou3519

**Comment by soumith** [comment]
Thanks @kevinzakka !
